pilot/pkg/serviceregistry/mcp/controller.go

```diff
@@ -125,6 +133,53 @@ func (c *controller) List(typ resource.GroupVersionKind, namespace string) (out
 	return out, nil
 }
 
+func (c *controller) Distribute(group string, s *snapshotter.Snapshot) {
+	switch group {
+	case snapshots.Istiod, snapshots.SyntheticServiceEntry:
+	default:
+		return
+	}
+
+	c.directMu.Lock()
+	defer c.directMu.Unlock()
+	for _, collection := range s.Collections() {
+		if _, exists := c.schemas.Find(collection); !exists {
+			continue
+		}
+
+		version := s.Version(collection)
+		if version == c.directVersion[collection] {
+			continue // skip unchanged collections
+		}
+
+		resources := s.ResourceInstances(collection)
+		objects := make([]*sink.Object, 0, len(resources))
+		for _, r := range resources {
+			metadata, err := resource2.SerializeMetadata(r.Metadata)
+			if err != nil {
+				scope.Processing.Errorf("Error serializing metadata for resource (%v): %v", r, err)
+				continue
+			}
+
+			objects = append(objects, &sink.Object{
+				Metadata: metadata,
+				Body:     r.Message,
+			})
+		}
+		change := &sink.Change{
+			Collection: collection,
+			Objects:    objects,
+		}
+		if err := c.Apply(change); err != nil {
+			log.Errorf("failed to apply collection=%v version=%v: err=%v",
```

Comment:
Just curious: what is the criteria for `scope.Processing.Errorf` vs `log.Errorf`?